iwyu_preprocessor.cc

```diff
@@ -1087,4 +1126,21 @@ bool IwyuPreprocessorInfo::ForwardDeclareIsInhibited(
       ContainsKey(*inhibited_forward_declares, normalized_symbol_name);
 }
 
+bool IwyuPreprocessorInfo::ForwardDeclareInKeepRange(
+    const clang::SourceLocation& loc) const {
+  // Check if the entry exists, shortcircuit if not
+  const clang::FileEntry* ref = GetFileEntry(loc);
+  if (keep_location_ranges_.find(ref) == keep_location_ranges_.end()) {
+    return false;
+  }
+
+  const clang::SourceRange mRange{loc};
+  for (auto const& range : keep_location_ranges_.at(ref)) {
```

Comment:
We've already done this lookup with the find above, so consider caching find result in an iterator:
```
auto it = keep_location_ranges_.find(ref);
if (it == keep_location_ranges_.end()) {
  return false;
}

for (const auto& range : *it) {
   ...
}